File: src/leetCodeProblems/HashMap/HappyNumber202.java
package leetCodeProblems.MathCalculations;

/**
 * LeetCode - https://leetcode.com/problems/happy-number/submissions/
 *
 * TimeComplexity - O(n)
 * SpaceComplexity - O(n)
 */

import java.util.HashSet;

public class 202 {

    private int getDigitsSum(int n) {

        int leftNumber = n;
        int digit = 0;
        int sum = 0;

        while(leftNumber > 0) {
            digit = leftNumber%10;
            sum += digit*digit;
            leftNumber = leftNumber/10;
        }

        return sum;
    }

    public boolean isHappy(int n) {

        HashSet sumAlreadySeen = new HashSet();

        while (n!=1 && !sumAlreadySeen.contains(n)) {
            sumAlreadySeen.add(n);
            n = getDigitsSum(n);
        }

        if (n == 1) {
            return true;
        }

        return false;
    }

    public static void main(String[] args) {

        202 obj = new HappyNumber202();

        System.out.println(obj.isHappy(2));
    }
}

File: src/leetCodeProblems/HashMap/HappyNumber202.go
package main

import "log"

/*
- Leetcode - https://leetcode.com/problems/happy-number/
- Time - O(nlogn)
- Space - O(n)
*/

func calculateSquaresSum(n int) int {
	sum := 0

	for n != 0 {
		r := n % 10
		sum += r * r

		n = n / 10
	}
	return sum
}
func isHappy(n int) bool {
	m := make(map[int]struct{})

	for n != 1 {
		if _, ok := m[n]; ok {
			return false
		}

		m[n] = struct{}{}
		n = calculateSquaresSum(n)
	}
	return n == 1
}

func main() {
	log.Println(isHappy(19))
	log.Println(isHappy(2))
	//log.Println(intToRoman(1994))
}
